### ISSUE TITLE
Add printer in LAN mode by IP address

### ISSUE THREAD

**Issue by rkarlsba** [bambulab/BambuStudio#702]
Since the WAN mode is [utterly insecure](https://blogg.karlsbakk.net/2022/11/23/bambu-lab-x1-carbon-the-flipside/), adding a printer by its IP address would be nice in LAN mode when mDNS isn't available. Either because the printer is on a different subnet and the router doesn't route multicast, which is quite normal or if you have an enterprise wifi system (for instance from Cisco), which also may drop mDNS.

**Comment by c2h5oh** [comment]
Multicast being blocked between wired and wireless connections is fairly common even on consumer/SOHO devices, so this would be nice even just for better connectivity.

**Comment by merasil** [comment]
+1
i have multiple vlans and adding the printer via ip would be awsome. mDNS is pain in the ass for subnets...

**Comment by gashton** [comment]
+1
Not sure why this wasn't done in the first place... definitely a must-have!
Bambu appear to have implemented Multicast SSDP instead of mDNS, Which is a shame as Avahi does not support this.

**Comment by rkarlsba** [comment]
From Wikipedia [SSDP] was formally described in an IETF Internet Draft by Microsoft and Hewlett-Packard in 1999. Although the IETF proposal has since expired (April, 2000),[1] SSDP was incorporated into the UPnP protocol stack, and a description of the final implementation is included in UPnP standards documents

Doesn't seem very sane to use with anything, tbh.

**Comment by c2h5oh** [comment]
Even fallback to regular DNS would be an improvement - I can always setup a local DNS entry for the printer..

**Comment by rkarlsba** [comment]
Which reminds me that IPv6 support would be nice. Seems the printer is running Linux after all, so it really can't be that hard.

**Comment by merasil** [comment]
> +1
> Not sure why this wasn't done in the first place... definitely a must-have!
> Bambu appear to have implemented Multicast SSDP instead of mDNS, Which is a shame as Avahi does not support this.

That explains why i cant get the Printer working in lan mode in my vlan network… i should have used wireshark in the first Place…

**Comment by rkarlsba** [comment]
> That explains why i cant get the Printer working in lan mode in my vlan network… i should have used wireshark in the first Place…

You can remedy that by enabling multicast routing or forwarding between VLANs if the router understands that sort of stuff. Still, adding by IP or DNS name shouldn't be very hard to fix and is the good solution to this.

**Comment by R1N4x** [comment]
Lots of fixes and improvements with v1.4 but this one is still missing.

**Comment by rkarlsba** [comment]
Can it really be that hard?

**Comment by rehashedsalt** [comment]
With today's outage, this is now especially pertinent. There is _already an existing dialogue for this feature_ as of 1.4.2 -- there should be very few blockers for this now.

**Comment by julie777** [comment]
My thoughts on this. 
It should not require an app using bluetooth to set up the printer for LAN mode. The selection for lan mode should always be available.
After selecting LAN mode only there needs to be a way to connect the printer to the network. A simple way would be to allow putting a file "network credentials.txt" on the SD card and the printer can use network name and password from that file to connect to the network.

When the printer connects to the network it should display its ip address and passcode for bambu studio to connect to it. I am not a network expert so I am not sure about broadcasting its availability on the network. I think this should be okay on a LAN, which makes it discoverable without having to type in the IP address in bambu studio.

**Comment by kriordan37** [comment]
+1 i should be able to use static ip addresses and point bambu studio to that. this issue is preventing me from using the printer in lan mode with a vpn

**Comment by tinuva** [comment]
With the latest issue happening now: https://www.reddit.com/r/BambuLab/comments/15rznpy/printer_started_printing_by_itself_damaged_itself/

This issue should really receive priority asap!

**Comment by ellenhp** [comment]
I posted this in another issue but it looks like this is the primary issue for adding printers by IP address so I'm going to post it here too, sorry about the spam. I can delete the comment in the other issue if it's a problem :)

I got a little bit frustrated with this issue because having command and control in the slicer is a key feature IMO. So I reimplemented the networking plugin to add this feature (and others).

https://github.com/ellenhp/bambu-farm

**Comment by timothyzalusky** [comment]
Another +1 for this feature.

**Comment by fritzw** [comment]
> As a temporary solution. I've written a [script](https://github.com/gashton/bambustudio_tools/blob/master/bambudiscovery.sh) to generate an unsolicited response to the upnp/ssdp discover, to tell BambuStudio the IP of the printer. It's meant to be run on Linux, but may work on Windows too if run in a VM/Container/Cygwin.
> 
> It has to be run each time BambuStudio is restarted, as Bambu will forget until it receives the response to its M-SEARCH ssdp:discover.

I've [converted the SSDP script to PowerShell](https://gist.github.com/fritzw/7d981046b509a27e204d01c9bd73f37b), so that it can be used on Windows as well.

**Comment by TonyAlberto** [comment]
> > As a temporary solution. I've written a [script](https://github.com/gashton/bambustudio_tools/blob/master/bambudiscovery.sh) to generate an unsolicited response to the upnp/ssdp discover, to tell BambuStudio the IP of the printer. It's meant to be run on Linux, but may work on Windows too if run in a VM/Container/Cygwin.
> > It has to be run each time BambuStudio is restarted, as Bambu will forget until it receives the response to its M-SEARCH ssdp:discover.
> 
> I've [converted the SSDP script to PowerShell](https://gist.github.com/fritzw/7d981046b509a27e204d01c9bd73f37b), so that it can be used on Windows as well.

I just tested it and it's ok, it works, thanks! I will only add that it may be necessary to change the port, in my case it is listening on port 1990 instead of 2021 (here is information about these ports: https://wiki.bambulab.com/en/general/printer-network-ports). The $TARGET_IP line itself is not needed in the powershell script because it is not used at all in this case.

**Comment by hoegge** [comment]
It is quite crazy you cannot add a printer by IP - it means a lot of user simply cannot connect to it and means you can return the printer and get a refund, since it does not work.

**Comment by ellenhp** [comment]
I imagine it's hard to find time to figure out the user experience, but if anyone from Bambu is reading this, I think I'd like to suggest that you add a way to configure printers by static IP in the config file, and worry about the slicer UI code for that configuration later when you have more time available. A lot of users would be okay editing a config file if it means not having to return the printer (or in my case, re-implement the entire networking plugin in Rust).

To anyone here who runs linux and likes fiddling around with semi-broken software, the [bambu-farm](https://github.com/ellenhp/bambu-farm) repo I linked earlier will get you up and running. It's not perfect. Immediately after a print it drops into an infinite loop on a worker thread so you have to close the slicer and restart the server-side program too, but it gets the job done for me and might be useful for someone else at their wit's end.

**Comment by RagingRoosevelt** [comment]
Yeah, without this, the X1E that my employer bought doesn't work on our corporate network. I can ping the printer and it's on the same subnet as my computer, but because it's trying to auto-discover, I can't add it to Bambu Studio.

**Comment by rehashedsalt** [comment]
>Yeah, without this, the X1E that my employer bought doesn't work on our corporate network.

Corporate environments are an especially pertinent example -- mDNS is oftentimes disabled as a basic security measure.

**Comment by RagingRoosevelt** [comment]
> mDNS is oftentimes disabled as a basic security measure

That's exactly what's causing the issues I'm running into.  And for Bambu to sell the printer specifically on the merits of compatibility with a corporate environment, but not support this direct IP / hostname specification seems disingenuous at best.

**Comment by walterwongbbl** [comment]
We are considering this feature, and recently it's the Chinese New Year holiday, so our responses might be slower. We apologize for any inconvenience caused.

**Comment by rkarlsba** [comment]
Happy new Chinese year! But I wonder - it's been more than 14 months since I posted this. Does it really take that long to fix such a minor thing?

**Comment by kluppy** [comment]
Without the ability to add printers via IP the software is not fit for purpose in a corporate environment.
I think I will be returning my X1E as such.

**Comment by Kablahz** [comment]
+1. Currently have an X1E in a corporate environment and have had to resort to FTPS to allow users to upload 3mf files from Studio. Company wants to purchase additional X1E units but we have put this on hold until we can statically address these printers and not rely on mDNS(not going to happen) or FTPS. Please make this a priority, should have been implemented before the X1E release.

**Comment by M1-Olsson** [comment]
So the X1E has the same problem, lol Enterprise edition with a extreme premium in price and still not working.

We could use another printer but I guess we stick to actual enterprise brands and not the hobby ones.

**Comment by rkarlsba** [comment]
> +1. Currently have an X1E in a corporate environment and have had to resort to FTPS to allow users to upload 3mf files from Studio. Company wants to purchase additional X1E units but we have put this on hold until we can statically address these printers and not rely on mDNS(not going to happen) or FTPS. Please make this a priority, should have been implemented before the X1E release.

As [gashton](https://github.com/gashton) wrote above, it's not even mDNS, it's SSDP, which isn't described in any valid IETF document, but later implemented in uPnP, but who on earth would allow uPnP on an enterprise network…

**Comment by ColColonCleaner** [comment]
I hope this is fixed soon. Just got an X1C and cannot use it over my network because of this issue.

**Comment by timothyzalusky** [comment]
Maybe we can light a fire under bambu lab by mentioning the competitors that have IP address printing...
Ultimaker
Makerbot
Creality
Formlabs
Markforged
I could go on.  We need this feature.

**Comment by rkarlsba** [comment]
Octoprint
Fluidd
Mainsail
(anything, really, except Bambu)

**Comment by ColColonCleaner** [comment]
It's really strange there is no movement or comment about this problem from Bambu. Not sure why they are seemingly ignoring it, the issue has been around for a long time now.

**Comment by rkarlsba** [comment]
They are passively ignoring this, for reasons unknown. Perhaps they make their money out of cloud traffic?

**Comment by M1-Olsson** [comment]
They have LAN mode so the could traffic is optional already. I use the script mentioned above and it works great so the problem is that the client can't find the printer itself unless on the same network. I can't understand why they just don't fix this, maybe we will have better luck with the Orca guys fixing this... 
I did see a windows once when trying to start the video stream asking for an IP but that seems only be for the video.

**Comment by ColColonCleaner** [comment]
> They have LAN mode so the could traffic is optional already. I use the script mentioned above and it works great so the problem is that the client can't find the printer itself unless on the same network. I can't understand why they just don't fix this, maybe we will have better luck with the Orca guys fixing this... 
> I did see a windows once when trying to start the video stream asking for an IP but that seems only be for the video.

The script doesn't work for me for some reason. Not sure why. Is there a trick to using it that I'm missing?
Printer never shows up.

The only way I can interact with the printer over the network is via FTP to drop gcode files in.

**Comment by rkarlsba** [comment]
I don't think you can fix this with orcaslicer either - I can only guess it's the bambu connectors that are doing this - otherwise, it'd be replaced long time ago. These are closed source, and AFAICT, breaks AGPL quite clearly. I wrote a piece about that a year and a half ago and got an answer from Bambu that wasn't very comforting, but I didn't follow up on it. I'll see what I can do in the future.

**Comment by ColColonCleaner** [comment]
@rkarlsba did you have to change that script at all to make it work, aside from filling in the printer attributes? It doesn't work for me.
I'll try running wireshark to see what port bambustudio is expecting and see if that's different, but otherwise I'm not sure why others are able to use it but it doesn't work for me.

**Comment by ellenhp** [comment]
> I can only guess it's the bambu connectors that are doing this - otherwise, it'd be replaced long time ago. These are closed source, and AFAICT, breaks AGPL quite clearly.

This is correct, but I reimplemented the networking plugin in a weekend or so for [bambu-farm](https://github.com/ellenhp/bambu-farm). This is very fixable.

**Comment by ellenhp** [comment]
Yeah I've dropped the link a few different times in case it's helpful to people, but like I mention in the readme it's a very rough piece of software, not at all something an average person can just pick up and use. Also only builds and runs on linux, but it might be a good starting point for someone who wants to fix this once and for all (probably in collaboration with orcaslicer)

**Comment by timothyzalusky** [comment]
This is in the top 10 based on number of comments, but if you add up each separate issue thread.  It is well into the 100s of comments on this issue.  @walterwongbbl please give us a timeline (I was expecting this resolved in the first year, not two years later).

**Comment by timothyzalusky** [comment]
Add printer in LAN mode by IP address #702
Printer IP Dialog #3714
LAN mode does not work between VLANs #3272
[Feature Request] Ability to add Printer by IP address (Lan Only Mode) #2426
Add device by IP address, without needing to trigger a connection error first #1549
Ability to enter IP of Printer(s) #2232
Fixed IP Bambulab X1 Carbon and P1P #1737

@walterwongbbl that is 7 separate issues (possibly more that I missed) that will be resolved by this.  Ulitmaker can do it so can you.

**Comment by Alex-Schaefer** [comment]
One could think they want all people, including companies, to use their cloud and upload everything there. Why, and what they are doing with the files, is up to you to assume.

Difficult to believe, that the ability to enter an IP address is such a difficult task.

**Comment by timothyzalusky** [comment]
They wouldn't have released the X1E that can be ran fully offline if that truly were their intent.  I would even pay more to have this capability.  Come on Bambu Lab!

**Comment by Alex-Schaefer** [comment]
X1C and other printers can use the LAN only mode, too. And what I read, the SSDP discovery problem is the same with X1E: If the printer can not be detected by BambuStudio/OrcaSlicer, you can't use it.

**Comment by M1-Olsson** [comment]
> X1C and other printers can use the LAN only mode, too. And what I read, the SSDP discovery problem is the same with X1E: If the printer can not be detected by BambuStudio/OrcaSlicer, you can't use it.

Correct, we use the scripts mentioned above as a workaround but that only makes this more comical as that solves the problem so the only thing missing is a way to enter the IP in the slicer and it will work....

**Comment by rkarlsba** [comment]
My printer is down atm, waiting for parts, but I just wonder - has anyone tried to use https://github.com/initstring/evil-ssdp to intercept and perhaps spoof the SSDP?

**Comment by elexx** [comment]
> My printer is down atm, waiting for parts, but I just wonder - has anyone tried to use https://github.com/initstring/evil-ssdp to intercept and perhaps spoof the SSDP?

Yes, see above :-)
* Bash script: https://github.com/bambulab/BambuStudio/issues/702#issuecomment-1347672299
* Powershell script: https://github.com/bambulab/BambuStudio/issues/702#issuecomment-1798553304

**Comment by Alex-Schaefer** [comment]
> My printer is down atm, waiting for parts, but I just wonder - has anyone tried to use https://github.com/initstring/evil-ssdp to intercept and perhaps spoof the SSDP?

I created a python script derived from the mentioned bash script, you can use it to send a fake SSDP packet with the printer IP from your PC to your PC 😬 
https://gist.github.com/Alex-Schaefer/72a9e2491a42da2ef99fb87601955cc3
It's linked in another similar issue, too: https://github.com/bambulab/BambuStudio/issues/2426#issuecomment-1700541962

**Comment by rkarlsba** [comment]
> As a temporary solution. I've written a [script](https://github.com/gashton/bambustudio_tools/blob/master/bambudiscovery.sh) to generate an unsolicited response to the upnp/ssdp discover, to tell BambuStudio the IP of the printer. It's meant to be run on Linux, but may work on Windows too if run in a VM/Container/Cygwin.
> 
> It has to be run each time BambuStudio is restarted, as Bambu will forget until it receives the response to its M-SEARCH ssdp:discover.

I made a few, tiny changes in a couple of PRs ;)

**Comment by rkarlsba** [comment]
Btw, It should be possible to use the same method in OrcaSlicer to just spoof the SSDP if needed. Has anyone looked into that?

**Comment by rkarlsba** [comment]
I thought about integrating something in orcaslicer to allow this without an external script

**Comment by Alex-Schaefer** [comment]
I thought about implementing this in OrcaSlicer myself, too. But I'm lacking the time for it and am not very familiar with C++.
And honestly I don't think this should be something the customer should do, this should be the responsibility of BambuLab.

**Comment by rkarlsba** [comment]
I cannot but agree, but if BambuLab doesn't care, which seems to be the case, it's better to have a good workaround than nothing.

**Comment by ellenhp** [comment]
> I thought about integrating something in orcaslicer to allow this without an external script

If you're ok modifying OrcaSlicer you don't need to do any of the SSDP spoofing stuff. All the networking plugin does is (through a couple of intermediate steps) call `on_machine_alive` when it receives the ssdp packet. Call that function directly with [appropriately formatted JSON](https://github.com/ellenhp/bambu-farm/blob/main/bambu-farm-client/src/api.rs#L168-L183) and you're good to go.

**Comment by rkarlsba** [comment]
Probably a five minute job if you know the code - I don't :P

**Comment by theGidy** [comment]
Thanks for the workarounds, i will trying them out. Still small hope for bambu to fix this issue.  First bambu is also the last bambu if this will not be fixed.

**Comment by tdgunes** [comment]
> We are considering this feature, and recently it's the Chinese New Year holiday, so our responses might be slower. We apologize for any inconvenience caused.

Hi there, are there any updates regarding this? @walterwongbbl @MackBambu

**Comment by ookurogullari** [comment]
Looking forward to having LAN Mode on my P1S! Is there any expected delivery date for it @walterwongbbl @MackBambu ?

**Comment by brbubba** [comment]
I see there's a flag for whether the printer is bound to a cloud account. I had to bind to an account for firmware update. Will the flag cause a refusal to connect or any other negative results? How do you unbind?

**Comment by rkarlsba** [comment]
I've tried to ask Bambu several times on how to upgrade firmware offline, as in downloading the firmware and just flashing it, as we have done with equipment for several decades, but they just reply, like a robot, NOT SUPPORTED.

**Comment by timothyzalusky** [comment]
I encourage everyone interested to go make a support ticket to request this, as well as user account management.  This is essential now that the have the X1E and want to be in the enterprise market.

**Comment by rkarlsba** [comment]
We also need offline firmware upgrades. I've talked to Bambu about it and all they can say is "not supported". Well, this method has been used for 30 years so it's no reason at all not to support it - they just want all machines into their domain.

**Comment by superbreadi** [comment]
I also require LAN mode because the printer is located in a separate VLAN to enhance data security. Currently, I am unable to access the printer via Bambu Studio, as it cannot be discovered through mDNS accross VLANs.

**Comment by rkarlsba** [comment]
It's not even mDNS, it's SSDP, which started out as a crude draft and then later became part of uPnP. But - since SSDP is based on UDP, you can spoof it
Python: https://gist.github.com/Alex-Schaefer/72a9e2491a42da2ef99fb87601955cc3
PowerShell: https://gist.github.com/fritzw/7d981046b509a27e204d01c9bd73f37b

**Comment by superbreadi** [comment]
However, does this mean I have to spoof it every time I start Studio? Additionally, I would like to use Bambu Handy via VPN in LAN mode. Is there a workaround for this as well?

**Comment by rkarlsba** [comment]
Btw, Bambu knowingly breaks AGPL with the bambu drivers https://blogg.karlsbakk.net/2024/05/24/more-on-bambus-license-issues/

**Comment by sprior** [comment]
If it helps I'm running my X1C in a different VLAN than my desktop machine, I use pfSense as my router/firewall and I installed the Avahi plugin on pfSense with the option to repeat packets across subnets.  With the appropriate ports opened between the subnets Bambu Studio works just fine.

**Comment by rkarlsba** [comment]
It's better to spoof it, so that you don't need to open up unnecessary ports. Also, Bambu really should fix this, along with allowing offline firmware updates.

**Comment by timothyzalusky** [comment]
> We also need offline firmware upgrades. I've talked to Bambu about it and all they can say is "not supported". Well, this method has been used for 30 years so it's no reason at all not to support it - they just want all machines into their domain.
[Look what was just announced, coming soon!](
https://forum.bambulab.com/t/x1e-now-supports-offline-firmware-update/89337)

**Comment by RagingRoosevelt** [comment]
Yeah, offline firmware is available as of recently:

https://wiki.bambulab.com/en/x1/manual/X1E-firmware-update-from-SD-card
https://www.youtube.com/watch?v=Gj7c5ANN-X0

You need to be on v01.01.02.00 which mean an online firmware upgrade first, but at least it's possible now.

**Comment by rkarlsba** [comment]
That doesn't help all of us that don't have an X1E. I have an X1C, which is basically the same machine as the X1E and this functionaliry has been, and still is, used for all sorts of devices

**Comment by ColColonCleaner** [comment]
A collaborator mentioned in #2232 that this will be added in the next update. I hope that actually happens, and am looking forward to it if it does.

**Comment by kluppy** [comment]
> That doesn't help all of us that don't have an X1E. I have an X1C, which is basically the same machine as the X1E and this functionaliry has been, and still is, used for all sorts of devices

Have you tried it? The title of the video embedded in the how to page is "Offline Firmware Update using SD Card - X1/X1C/X1E"
I assume they all run the same firmware really.

**Comment by mwdle** [comment]
> > As a temporary solution. I've written a [script](https://github.com/gashton/bambustudio_tools/blob/master/bambudiscovery.sh) to generate an unsolicited response to the upnp/ssdp discover, to tell BambuStudio the IP of the printer. It's meant to be run on Linux, but may work on Windows too if run in a VM/Container/Cygwin.
> > It has to be run each time BambuStudio is restarted, as Bambu will forget until it receives the response to its M-SEARCH ssdp:discover.
> 
> I've [converted the SSDP script to PowerShell](https://gist.github.com/fritzw/7d981046b509a27e204d01c9bd73f37b), so that it can be used on Windows as well.

Can confirm this works perfectly for remotely accesing my BambuLab A1 (with a few minor modifications like SN/Model) over my Wireguard VPN (Tailscale).

**Comment by tdgunes** [comment]
Sorry for multiple messages however, are there any updates regarding this? @walterwongbbl @MackBambu

**Comment by kluppy** [comment]
@tdgunes It has been flagged for the 1.10 release.
https://github.com/bambulab/BambuStudio/issues/4512#issuecomment-2365472604

**Comment by tdgunes** [comment]
> @tdgunes It has been flagged for the 1.10 release. [#2232 (comment)](https://github.com/bambulab/BambuStudio/issues/2232#issuecomment-2310529584) [#4512 (comment)](https://github.com/bambulab/BambuStudio/issues/4512#issuecomment-2365472604)

Thanks, however it is not clear to me if this feature will be for X1E only, or for all BambuLab printers. Do you have any further information @kluppy?

**Comment by ColColonCleaner** [comment]
@tdgunes the people who posted these issues are using X1C/P1S printers so it should apply to at least those and also the X1E.

**Comment by tdgunes** [comment]
> @tdgunes the people who posted these issues are using X1C/P1S printers so it should apply to at least those and also the X1E.

I see, however I would like this feature to be implemented for A1 series, as the LAN mode does not work effectively with those as well.

**Comment by psychedelicu** [comment]
Ia this available on P1S? Because I want to buy it only if it has this implemented already.

**Comment by kluppy** [comment]
@psychedelicu 1.10 has not been released yet, so no.

**Comment by psychedelicu** [comment]
> @psychedelicu 1.10 has not been released yet, so no.
> 

But they announce that this will pe included?

**Comment by MrSuicideParrot** [comment]
The public beta of 1.10 is available and includes binding by IP address!

**Comment by mwdle** [comment]
I'm just getting around to testing out the beta and I can't seem to get my printer to connect via IP address. I have a BambuLab A1.

![image](https://github.com/user-attachments/assets/a895f0b0-8250-4146-bca0-f1ed82d1d21a)

I've confirmed my access code and IP address are correct and I'm able to connect on LAN as one normally would.

Not sure if this is a bug or what but it would be interesting to know if anyone else is facing this issue.

**Comment by MrSuicideParrot** [comment]
I also tried and found the same error than you. 
I used wireshark  to understand what the application was doing, and it attempts to do a tcp connection with the port 3000  and receives a tcp reset from the printer.  Normally, the slicer talks with port 8883. So, this new features could depend on a firmware update  or they forgot to put the correct port. Either way, a firmware update shouldn't be needed for this to work.

**Comment by ColColonCleaner** [comment]
> the printer shouldn't need a firmware update for this to work right? I am under the impression that this feature is all implemented within Bambu Studio.

This should not require a firmware update on the printer. We can already get the slicer to talk with the printer by spoofing the response to studio's request so it definitely should not need any update on the printer side.

If they make it require a firmware update I'd wonder what else they changed.

**Comment by Rainie-Read-wdc** [comment]
It appears this feature in 1.10.0 is also broken or doesn't support my X1E. I am still able to use my wireshark spoofed response on the new version though. 
![image](https://github.com/user-attachments/assets/94437107-d44b-40e2-a842-d9c3d1ca3f6f)

**Comment by chrellrich** [comment]
We have bought two X1E at work and I get the same error when testing the feature in the beta release. Using a spoofed message it works.

**Comment by therealmrfox** [comment]
I was very happy to see this feature added, but no dice for me either. Though I get a slightly different error, "Failed to publish login request". (Occasionally Bambu Studio also decides to crash instead.) 2 steps forward, 1 step back...

Oddly this is [described](https://wiki.bambulab.com/en/bambu-studio/troubleshoot/login-binding-issue) as a failure to connect to Bambu cloud, which is of course nonsense when connecting in LAN mode.

P1S, works fine once the printer is discovered via (spoofed or real) SSDP, so definitely no networking issues here.

![Screenshot from 2024-10-23 20-49-14](https://github.com/user-attachments/assets/2e3c13f1-25b9-474f-8f6c-e7daa892fb66)

Filed [issue 5070](https://github.com/bambulab/BambuStudio/issues/5070).

**Comment by AndreasSchwalb** [comment]
Maybe anybody is interested in:
I wrote a small relay which take the SSDP messages the printer send over the network and send it to the machine where BambuStudio is running on.
For my case it is a nice workaround. Just spin up the container and BambuStudio is happy.

https://github.com/AndreasSchwalb/babulabRelay

**Comment by hofbauerc** [comment]
In my case this feature still does not work. I type in the IP and Access code. It displays "connecting to printer". In the next window I type in SN it displays "connecting to printer, this dialog will close later".

After that, nothing happens. I can do it again but still nothing changes.

Does this feature work for anyone? In the previous BambuStudio versions I triggered Bambu studio with the IP of my system via a custom python script which worked without issues.

**Comment by pelaxa** [comment]
> I cannot test this on my X1C until new firmware is available.
> 
> From the [Studio 1.10.0.89 public release notes](https://wiki.bambulab.com/en/software/bambu-studio/release/release-note-1-10-0): Subnet binding support: Users can now bind printers across different subnets by directly entering the printer's IP address and Access Code (Supported P-series firmware version: 01.07.00.00, supported A-series firmware version: 01.04.00.00, _**supported X-series firmware: next release version).**_

Looks like the P-Series firmware is not out either

**Comment by hofbauerc** [comment]
> I cannot test this on my X1C until new firmware is available.
> 
> From the [Studio 1.10.0.89 public release notes](https://wiki.bambulab.com/en/software/bambu-studio/release/release-note-1-10-0): Subnet binding support: Users can now bind printers across different subnets by directly entering the printer's IP address and Access Code (Supported P-series firmware version: 01.07.00.00, supported A-series firmware version: 01.04.00.00, _**supported X-series firmware: next release version).**_

Ah okay so the X1C does not have the appropriate firmware that makes sense why it does not work currently.

**Comment by juvinious** [comment]
> Maybe anybody is interested in: I wrote a small relay which take the SSDP messages the printer send over the network and send it to the machine where BambuStudio is running on. For my case it is a nice workaround. Just spin up the container and BambuStudio is happy.
> 
> https://github.com/AndreasSchwalb/babulabRelay

Seems to work up until it adds the printer. I get the prompts after entering IP and Access Code for the serial number but it just does nothing without any error. The popup GUI just goes away as if it's going to add the printer, but it is never added.

**Comment by AndreasSchwalb** [comment]
> > Maybe anybody is interested in: I wrote a small relay which take the SSDP messages the printer send over the network and send it to the machine where BambuStudio is running on. For my case it is a nice workaround. Just spin up the container and BambuStudio is happy.
> > https://github.com/AndreasSchwalb/babulabRelay
> 
> Seems to work up until it adds the printer. I get the prompts after entering IP and Access Code for the serial number but it just does nothing without any error. The popup GUI just goes away as if it's going to add the printer, but it is never added.

You don't need to enter an IP address in the BambuStudio GUI for this "hack". You just need to start the container with the right configuration. But it is a little bit off-topic of this issue. If you need any further advice you may open an issue in the bambulabRelay Repository

**Comment by juvinious** [comment]
> > > Maybe anybody is interested in: I wrote a small relay which take the SSDP messages the printer send over the network and send it to the machine where BambuStudio is running on. For my case it is a nice workaround. Just spin up the container and BambuStudio is happy.
> > > https://github.com/AndreasSchwalb/babulabRelay
> > 
> > 
> > Seems to work up until it adds the printer. I get the prompts after entering IP and Access Code for the serial number but it just does nothing without any error. The popup GUI just goes away as if it's going to add the printer, but it is never added.
> 
> You don't need to enter an IP address in the BambuStudio GUI for this "hack". You just need to start the container with the right configuration. But it is a little bit off-topic of this issue. If you need any further advice you may open an issue in the bambulabRelay Repository

Will do, let me play around with it if I can't get it to work, I'll open a ticket. Thanks!

**Comment by Alex-Schaefer** [comment]
@juvinious 
If you're using external tools anyway, I may suggest you look into my "fake SSDP script": https://gist.github.com/Alex-Schaefer/72a9e2491a42da2ef99fb87601955cc3
Can use hardcoded values or be used with arguments. The only thing it does is creating an SSDP packet with the provided printer information and sends it to localhost (or any other provided IP). Run it once after BambuStudio/OrcaSlicer is open.
Or if you don't have Python installed anyway, you may look into a PowerShell version af a similar script: https://github.com/withanhdammit/BambuStudio_Launcher/
Might be a bit easier to understand.

-----

On topic: Honestly, it's ridiculous that this new feature needs a firmware update. They could have implemented this in a way that would have worked without any firmware update.
Why they have chosen to make life complicated for both themselves and the users, I can't understand.

**Comment by jonidesousa** [comment]
anything for macos users?

**Comment by koniiiik** [comment]
> In my case this feature still does not work. I type in the IP and Access code. It displays "connecting to printer". In the next window I type in SN it displays "connecting to printer, this dialog will close later".
> 
> After that, nothing happens. I can do it again but still nothing changes.

I've been observing the same behavior. After entering all four of the required fields, I get the message “connecting to printer, this dialog will close later”, and after a few seconds, the dialog disappears without anything changing in the main UI.

**Comment by koniiiik** [comment]
> Which printer and which firmware version did you try?
> Did you recognize this post?
> [#702 (comment)](https://github.com/bambulab/BambuStudio/issues/702#issuecomment-2473433031)

The printer is A1 mini, firmware 01.03.01.02 which is the latest at the moment according to the [release history on the bambulab website](https://wiki.bambulab.com/en/a1-mini/manual/a1-mini-firmware-release-history). Bambu studio version 1.10.1.50, which, again, as far as I can tell is the latest at the moment.

As for the comment that you're referencing, are you implying that the feature is not actually usable with any released firmware as of right now? When I check the release history for A1 and P1 firmware, neither lists a version matching what's listed as required in the release notes.

Weirdly enough, I tried again, and suddenly it worked. I didn't do anything different, it just started to work out of nowhere. Which also seems to disprove the firmware version requirements.

I have no idea what's going on here, I'm just glad that I got the thing to work. A lot of the information presented here and in official release notes doesn't seem to match with reality. :shrug:

**Comment by narner90** [comment]
Here is a python script to advertise a printer that is tested working on macos with a P1S on a separate VLAN:

[https://gist.github.com/narner90/4c6cd23c452df990606e9b848938b8ee](https://gist.github.com/narner90/4c6cd23c452df990606e9b848938b8ee)

**Comment by Alex-Schaefer** [comment]
@narner90 with [socket](https://docs.python.org/3/library/socket.html) you shouldn't need netcat for this, see: https://gist.github.com/Alex-Schaefer/72a9e2491a42da2ef99fb87601955cc3

**Comment by gentlecat** [comment]
Have an A1 Mini. Even with firmware 01.04.00.00 and 1.10.1.50 version of the Bambu Studio, I cannot get it to work.

I've been using this tool so far: https://github.com/Rdiger-36/StudioBridge. But the whole thing is annoying.

**Comment by Meth0d007** [comment]
Whatever the "E" in 1XE stands for, it is not "Enterprise" since we cannot use our 1XE machines in properly segmented corporate network in 2025. Way to go Bambulab.

**Comment by Sector14** [comment]
Tried connecting via IP/access code on version 1.10.1.50 and similar problem to those above. Just get back the error message "failed to publish login request".

As it happens my PC and printer are on the same network (wifi/wired bridged), so I unblocked port 2021/udp and the printer was auto discovered. However, IP mode really needs to work because not all wifi networks are bridged with the lan.

**Comment by rkarlsba** [comment]
> Whatever the "E" in 1XE stands for, it is not "Enterprise" since we cannot use our 1XE machines in properly segmented corporate network in 2025. Way to go Bambulab.

I beleive the E is for "Extended price" ;)

**Comment by spigaudio** [comment]
I'm also still waiting for a firmware update for the X1E to work with the "connect by IP" feature. The Bambulab support cannot tell me if or when this update will come. 
The most expensive printer has the worst firmware support. 

I can only agree with that: E stands for "Extended price"!

**Comment by LJKraus** [comment]
FWIW, Studio 2.1.0.59 was released today, but does not seem to have corrected this issue.

You might want to try OrcaSlicer.  

"Connect by IP" in OrcaSlicer works well with the X1C and unlike Studio, OrcaSlicer remembers the LAN-connected printer without the need to re-select it for each session.  Supports sub-nets also.

**Comment by bcarrington** [comment]
For Mac users, I had Mr. Claude write something up to do the same thing and send SSDP over to my printer VLAN. It's working great, you can either run this Python script before opening Bambu Slicer, or you can set it up in LaunchAgent to have it run in the background. My printer shows up after a couple seconds in the slicer now instead of a couple minutes. It's digital duct tape but it'll do.

```
#!/usr/bin/env python3
"""
bambu_announce_loop.py

Repeatedly sends a fake SSDP announcement for your printer to Bambu
Studio's local listener (UDP port 2021 on localhost), so it shows up
in the device list within seconds instead of waiting on the printer's
own broadcast timer. Useful if your printer is on a different
VLAN/subnet than the machine running Bambu Studio.

Note: this only solves discovery — Bambu Studio still needs a normal
unicast network path to the printer's IP for everything after that.

Fill in the four values below, then run: python3 bambu_announce_loop.py

Model strings (see BambuStudio's resources/printers directory):
    X1C: "3DPrinter-X1-Carbon"   X1: "3DPrinter-X1"   X1E: "C13"
    P1P: "C11"   P1S: "C12"   A1 mini: "N1"   A1: "N2S"
"""

import socket
import time
from datetime import datetime, timezone

# ---- EDIT THESE THREE VALUES ----
PRINTER_IP = "192.168.1.100"        # your printer's actual IP address
PRINTER_SERIAL = "YOUR_SERIAL_HERE"  # printer's serial number
PRINTER_MODEL = "3DPrinter-X1-Carbon"  # see model strings above
PRINTER_NAME = "MyPrinter"          # any label — shown in Bambu Studio's device list
# ----------------------------------

TARGET_IP = "127.0.0.1"   # Bambu Studio always listens on localhost
TARGET_PORT = 2021
ANNOUNCE_INTERVAL_SECONDS = 5

DEV_SIGNAL = "-44"      # artificial signal strength; keeps the wifi icon full-bars
DEV_CONNECT = "lan"     # LAN mode
DEV_BIND = "free"       # not bound to a cloud account


def build_message() -> bytes:
    date_str = datetime.now(timezone.utc).strftime("%a, %d %b %Y %H:%M:%S GMT")
    msg = (
        "HTTP/1.1 200 OK\r\n"
        "Server: Buildroot/2018.02-rc3 UPnP/1.0 ssdpd/1.8\r\n"
        f"Date: {date_str}\r\n"
        f"Location: {PRINTER_IP}\r\n"
        "ST: urn:bambulab-com:device:3dprinter:1\r\n"
        "EXT:\r\n"
        f"USN: {PRINTER_SERIAL}\r\n"
        "Cache-Control: max-age=1800\r\n"
        f"DevModel.bambu.com: {PRINTER_MODEL}\r\n"
        f"DevName.bambu.com: {PRINTER_NAME}\r\n"
        f"DevSignal.bambu.com: {DEV_SIGNAL}\r\n"
        f"DevConnect.bambu.com: {DEV_CONNECT}\r\n"
        f"DevBind.bambu.com: {DEV_BIND}\r\n"
        "\r\n"
    )
    return msg.encode("ascii")


def main():
    print(f"Announcing '{PRINTER_NAME}' ({PRINTER_IP}) to Bambu Studio every "
          f"{ANNOUNCE_INTERVAL_SECONDS}s. Press Ctrl+C to stop.")
    message = build_message()
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        try:
            while True:
                sock.sendto(message, (TARGET_IP, TARGET_PORT))
                print(f"  sent announce for {PRINTER_NAME} ({PRINTER_IP})")
                time.sleep(ANNOUNCE_INTERVAL_SECONDS)
        except KeyboardInterrupt:
            print("\nStopped.")


if __name__ == "__main__":
    main()
```

**Comment by tfcoulson** [comment]
I have connect by IP working now over different vlans on a H2D. Firewall ports opened: 3002, 8883, 990, 50000-50100, 322 and 6000 TCP. Printer set to LAN mode and developer mode enabled (key part - this disables the firewall feature on the printer). I have been told this will be fixed in the next firmware release.
